### ISSUE TITLE
W1b: `project_points` — five projection entry points, four different answers at `z = 0`

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4267]
`kornia.geometry.camera` has **five** entry points that divide by the camera-frame `z`, and at the singular input `(x, y, z) = (1, 2, 0)` they return **four different answers**. Away from the singularity all five agree exactly.

| site | policy | answer at `z = 0` |
|---|---|---|
| `project_points` | masked divide (`abs(z) <= 1e-8` → skip the divide), K applied **after** | `[[104.0, 203.0]]` |
| `PinholeCamera.project` | the same mask, K applied **before** | `[[100.0, 200.0]]` |
| `project_points_z1` | plain divide | `[[inf, inf]]` |
| `Z1Projection.project` (`kornia.sensors`) | plain divide | `[[inf, inf]]` |
| `cam2pixel` | `x / (z + 1e-12)` | `[[[[1.0e14, 2.0e14]]]]` (`inf` in float16) |

Two of these are worth separating out. `PinholeCamera.project` and `project_points` are documented as the same projection and are the container/free-function pair; they apply `K` on opposite sides of the masked divide, so they **disagree exactly on the singular branch** — the divergence is the defect, not either value. And `cam2pixel`'s `+ 1e-12` is not a guard: it enters the arithmetic, dominates the result below `z ≈ 1e-10`, and biases every finite output (relative error `1e-06` already at `z = 1e-6`).

Separately, all five project points **behind** the camera silently: `z = -4` gives a finite pixel with no warning anywhere in the family.

## Repro

```python
import torch
from kornia.geometry.camera import PinholeCamera, cam2pixel, project_points, project_points_z1
from kornia.sensors.camera.projection_model import Z1Projection
from kornia.geometry.vector import Vector3

K4 = torch.eye(4)[None]; K4[0, 0, 0] = K4[0, 1, 1] = 100.0; K4[0, 0, 2], K4[0, 1, 2] = 4.0, 3.0
K3 = K4[:, :3, :3].contiguous()
cam = PinholeCamera(K4.clone(), torch.eye(4)[None], torch.tensor([6]), torch.tensor([8]))
p = torch.tensor([[1.0, 2.0, 0.0]])          # z = 0 exactly: the singular input

print("project_points          :", project_points(p, K3).tolist())
print("PinholeCamera.project   :", cam.project(p).tolist())
print("project_points_z1       :", project_points_z1(p).tolist())
print("Z1Projection.project    :", Z1Projection().project(Vector3(p)).data.tolist())
print("cam2pixel               :", cam2pixel(p[None, None], K4).tolist())
print("cam2pixel float16       :", cam2pixel(p[None, None].half(), K4.half()).tolist())
print("-- away from the singularity, z = 4 --")
q = torch.tensor([[1.0, 2.0, 4.0]])
print("project_points          :", project_points(q, K3).tolist())
print("PinholeCamera.project   :", cam.project(q).tolist())
print("cam2pixel               :", cam2pixel(q[None, None], K4).tolist())
print("-- a point behind the camera (z = -4) is projected silently --")
print("project_points(z=-4)    :", project_points(torch.tensor([[1.0, 2.0, -4.0]]), K3).tolist())
```

## Observed

```
project_points          : [[104.0, 203.0]]
PinholeCamera.project   : [[100.0, 200.0]]
project_points_z1       : [[inf, inf]]
Z1Projection.project    : [[inf, inf]]
cam2pixel               : [[[[100000000376832.0, 200000000753664.0]]]]
cam2pixel float16       : [[[[inf, inf]]]]
-- away from the singularity, z = 4 --
project_points          : [[29.0, 53.0]]
PinholeCamera.project   : [[29.0, 53.0]]
cam2pixel               : [[[[29.0, 53.0]]]]
-- a point behind the camera (z = -4) is projected silently --
project_points(z=-4)    : [[-21.0, -47.0]]
```

## Expected

One singular-input policy for the namespace, chosen once and applied to all five sites. The three coherent policies are:

1. **Return `inf`** (what `project_points_z1` and `Z1Projection.project` do): the mathematically honest answer, and the one a downstream `isfinite` check can act on.
2. **Raise** on a singular `z`: `project_points_z1` already *documents* the `z > 0` precondition without validating it (idiom 5), so this makes the docstring true.
3. **Keep the mask** but make `project_points` and `PinholeCamera.project` agree on which side of it `K` is applied.

`cam2pixel`'s `z + 1e-12` should not survive any of the three: an epsilon that perturbs a well-conditioned result is idiom 1, and its float16 behaviour (`inf` anyway) shows it does not even achieve what it was added for.

Whether points behind the camera should raise, warn, or stay silent is part of the same decision.

## Classification

**Semantic contract** — a convention conflict recorded as such, not a bug in any one function. Idioms 1 (`cam2pixel`'s `+ 1e-12`) and 4 (a singular input silently producing garbage rather than raising).

## Disposition

**KEEP SEPARATE — intentional variant** for the five symbols: they take different parametrizations (`(*, 3)` + `K`, `(*, 3)` normalized, `Vector3`, `(B, 3, H, W)`) and different input domains, so they are not alias-able onto one another. What is recorded here is the **conflict**: choosing one singular-input policy for the namespace is an idiom-1/idiom-4 sweep for the coordinated Repair & Deprecation window, not a merge. Batch 5 documents each site's current policy on its own symbol, with a pointer from each to this issue, and pins the four answers.

## Environment

kornia 0.9.0rc1 @ `92e5fe8d` (main), torch 2.14.0, python 3.11, cpu. Found during the camera conventions audit (batch 5, tracking issue #4262); the batch documents the behavior below as it is and pins it, and does not change it.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
One more instance, at the warp level, now documented on `DepthWarper` and pinned in #4317: `warp_frame_depth` projects through `project_points` (masked divide, skipped at `abs(z) <= 1e-8`) and `DepthWarper` through `cam2pixel` (`z + 1e-12`), so the two split outright when the transformed points reach `z = 0`. With `K = I`, unit depth and a `(+1, 0, -1)` translation, `warp_frame_depth(arange(1, 7).view(1, 1, 2, 3), ...)` returns `[[2, 3, 0], [5, 6, 0]]` — the image sampled at the undivided `(u + 1, v)` — while `DepthWarper` sends every pixel to a coordinate of order `1e12` (`inf` in float16, where the `1e-12` rounds to zero) and what comes back is backend-dependent: zeros on cpu float32/float64/bfloat16, `nan` on cpu float16, `1e24` in one row on mps float32. One half unit closer (`z = 0.5`) the two agree to `0.0` / `2.4e-11`.

Pin: `test_wart_warp_frame_depth_and_depth_warper_split_at_zero_transformed_depth_4267` in `tests/geometry/test_depth.py`; delete when a single `z = 0` policy lands, and note that any policy other than "keep the mask" changes `warp_frame_depth`'s output on this input.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).

**Comment by ducha-aiki** [comment]
A consequence of the `cam2pixel` `z + 1e-12` policy that goes beyond a wrong value: on torch 2.5.1 (kornia's floor) with the aarch64 CPU build, `DepthWarper` on points whose transformed depth reaches `z = 0` **segfaults the process**. The `1e12` pixel coordinate becomes a normalized grid coordinate above `2**31` on one axis while the other axis stays in range, and torch 2.5.1's vectorized `grid_sample` CPU kernel overflows int32 there (crash at `x = 2147483648.0, y = -1.0`; fine at `x = 2147483000.0`, at `x = -1e12`, or with both axes out of range). Upstream: pytorch/pytorch#24823. Reproduced on an arm64 Mac and on the macOS torch 2.5.1 legs of the #4317 CI at `c77ec04c`; x86 2.5.1 and torch ≥ 2.9.1 return zeros instead. The masked divide in `project_points` (what `warp_frame_depth` uses) never produces such a coordinate.

The #4317 pin from my previous note now asserts the grid coordinate only and no longer samples through it (`2bd698d5`); the rest of that note stands.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).

**Comment by ducha-aiki** [comment]
## Decision: one coordinate policy, with validity split out

Maintainer decision (@ducha-aiki). Of the three policies listed above, kornia adopts **option 3, keep the mask**, for every projection entry point in `kornia.geometry.camera` and `kornia.sensors.camera`. The validity question is split out to #4555.

**The rule.** Where `abs(z) <= eps` (`eps = 1e-8`, absolute, in the caller's camera units), the homogeneous divide is skipped and the undivided `(x, y)` is used. The intrinsics are then applied, exactly as `project_points` does today. Everywhere else the divide is exact, with no epsilon added to it. The result is a finite **placeholder**, not a valid pixel; whether a projection is valid is #4555's question.

Why this and not `inf` or raising:
- **It is the only one of the three that is finite, gradient-safe, `torch.compile`-safe and crash-free on our torch floor.** On torch 2.5.1 (arm64 CPU), `grid_sample` segfaults on an `inf` grid coordinate, just as it does on `1e12`, so `inf` would crash the depth warps.
- **Raising is not viable for depth maps.** Missing depth is encoded as `0`, which makes the singular case routine input rather than an error. Validating it would also need a data-dependent host sync.
- **It changes the least.** `project_points` and `warp_frame_depth` already behave this way.

**Requirements for the implementing PR** (#4475):
1. **NaN propagates.** Select the divided branch with `~(torch.abs(z) <= eps)`, not `torch.abs(z) > eps`. Every comparison with `nan` is `False`, so `abs(z) > eps` routes `nan` depth into the pass-through and returns a finite pixel. The negated form routes it into the divide, where it stays `nan`.
2. **One intrinsics ordering everywhere.** The mask comes first and `K` second, in `project_points`, `PinholeCamera.project` and `cam2pixel` alike. `cam2pixel`'s `z + 1e-12` goes away. The container and the free function, and the two depth warps (`warp_frame_depth` via `project_points`, `DepthWarper` via `cam2pixel`), must agree at `z = 0` with a nonzero principal point too.
3. **Angle-based models do not go through the Z1 pass-through.** `KannalaBrandtK3` computes `θ = atan2(‖(x, y)‖, z)` from the camera-frame ray, not `atan` of the Z1-projected radius. That makes `z = 0` exactly θ = 90° and keeps rays beyond 90° correct, instead of sending near-singular rays to the 45° ring.
4. **Points behind the camera keep today's projection arithmetic.** The docs say so, and flagging them is part of #4555.
5. **Gradients:** invalid-branch safety uses a safe-argument `torch.where`, per `AGENTS.md`, and is checked on torch 2.5.1.
6. **The `_4267` pins are replaced by convention tests in the same PR:**
   - `test_wart_project_points_skips_the_divide_at_z_zero_4267` (`test_perspective.py`)
   - `test_wart_cam2pixel_epsilon_makes_the_singular_divide_finite_4267` and `test_wart_project_and_project_points_disagree_at_z_zero_4267` (`test_pinhole.py`)
   - `test_wart_project_points_z1_zero_depth_is_component_dependent_4267` (`test_projections.py`)
   - `test_wart_warp_frame_depth_and_depth_warper_split_at_zero_transformed_depth_4267` (`test_depth.py`)
   - `test_wart_projection_differs_from_geometry_at_small_depth_4267` (`test_camera_model.py`)
   - `test_wart_project_divides_by_z_with_no_guard_4267` (`test_projection_model.py`)

   The `#4267` prose at `perspective.py:46`, `pinhole.py:401` and `:874`, `projection_z1.py:57`, `depth.py:538`, `camera_model.py:147` and `projection_model.py:47` is rewritten to state the rule. It is a breaking change with a `breaking` fragment.

This issue closes when that PR lands. Validity masks, what warps sample at invalid pixels, and the behind-camera policy continue in #4555.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
A sixth site for this issue: `reproject_disparity_to_3D` (and `StereoCamera.reproject_disparity_to_3D`) divides by the homogeneous `W = -fy * disparity` through `convert_points_from_homogeneous`, whose `abs(W) <= 1e-8` mask returns the numerator unchanged. A zero-disparity pixel, which is a point at infinity, therefore comes back as a finite point, and on a real rig that point is behind the camera:

```python
import torch
from kornia.geometry.camera.stereo import StereoCamera, reproject_disparity_to_3D

left = torch.tensor([[[100.0, 0, 4, 0], [0, 100, 3, 0], [0, 0, 1, 0]]]); right = left.clone(); right[0, 0, 3] = -50.0
cam = StereoCamera(left, right)
print(cam.reproject_disparity_to_3D(torch.zeros(1, 1, 1, 1)).tolist())   # [[[[200.0, 150.0, -5000.0]]]]
print(cam.reproject_disparity_to_3D(torch.full((1, 1, 1, 1), 10.0)).tolist())  # [[[[-0.2, -0.15, 5.0]]]]

q = torch.tensor([[[1.0, 0, 0, 1], [0, 1, 0, 2], [0, 0, 0, 3], [0, 0, 1, 0]]])
d = torch.zeros(1, 1, 1, 1)
print(reproject_disparity_to_3D(d, q).tolist(), reproject_disparity_to_3D(d, -q).tolist())  # [1, 2, 3] vs [-1, -2, -3]
```

Because the masked branch skips the divide, an overall sign on `Q` no longer cancels there: `Q` and `-Q` return opposite points for the same zero disparity, while they agree everywhere else. `cv2.reprojectImageTo3D` divides by `W` (giving `inf`), or with `handleMissingValues=True` maps the minimum disparity to `Z = 10000`. Whatever singular-input policy this issue settles should also cover this call. It is now documented as a wart on `StereoCamera` in #4790 and pinned as `test_wart_reproject_disparity_zero_disparity_returns_the_numerator_4267`.

kornia `main` @ 57cd68e6f, torch 2.14.0, CPU.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by ducha-aiki** [comment]
Correction to my "sixth site" comment above ([5796074380](https://github.com/kornia/kornia/issues/4267#issuecomment-5796074380)). It overlooked the decision posted earlier in this thread ([5678077316](https://github.com/kornia/kornia/issues/4267#issuecomment-5678077316)). That decision already settles `reproject_disparity_to_3D`.

The adopted rule is **keep the mask**: at `abs(z) <= eps` the divide is skipped, and the result is a finite placeholder, not a valid pixel. `reproject_disparity_to_3D` already follows that rule. At zero disparity it returns the numerator through the same `convert_points_from_homogeneous` mask, so the #4267 implementation has nothing to change there.

What is left for that pixel is a validity question, and validity belongs to #4555:
- the placeholder lies behind the camera on a real rig;
- `Q` and `-Q` disagree on it;
- nothing flags it as invalid.

#4790 now links the stereo docstring to #4555 and renames the pin to `test_wart_reproject_disparity_zero_disparity_returns_the_numerator_4555`. It also reframes the `project_points` warning. That warning no longer describes the adopted rule as the defect. The defect it describes is that the other entry points do not follow the rule yet.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).
